sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala

```diff
@@ -736,15 +691,156 @@ private[execution] final class UnboundedFollowingWindowFunctionFrame(
+ *
+ * This implementation only supports evaluation in [[Complete]] mode. This is enough for
+ * Window processing.
+ *
+ * Processing of distinct aggregates is currently not supported.
+ *
+ * The implementation is split into an object which takes care of construction, and a the actual
+ * processor class. Construction might be expensive and could be separated into a 'driver' and a
+ * 'executor' part.
+ */
+private[execution] object AggregateProcessor {
+  def apply(functions: Array[Expression],
+    ordinal: Int,
+    inputAttributes: Seq[Attribute],
+    newMutableProjection: (Seq[Expression], Seq[Attribute]) => () => MutableProjection):
+    AggregateProcessor = {
+    val aggBufferAttributes = mutable.Buffer.empty[AttributeReference]
+    val initialValues = mutable.Buffer.empty[Expression]
+    val updateExpressions = mutable.Buffer.empty[Expression]
+    val evaluateExpressions = mutable.Buffer.fill[Expression](ordinal)(NoOp)
+    val imperatives = mutable.Buffer.empty[ImperativeAggregate]
+
+    // Create and add a size reference to SizeBasedWindowFunction.
+    var sizeOrdinal = -1
+    var size: BoundReference = null
+    val addSize = (f: Expression) => f match {
+      case wf: SizeBasedWindowFunction =>
+        if (size == null) {
+          sizeOrdinal = aggBufferAttributes.size
+          size = BoundReference(sizeOrdinal, IntegerType, false)
+          aggBufferAttributes += wf.size
+          initialValues += NoOp
+          updateExpressions += NoOp
+        }
+        wf.withSize(size)
+      case e => e
+    }
+
+    // Add an AggregateFunction to the AggregateProcessor.
+    val addToProcessor = (f: Expression) => f match {
+      case agg: DeclarativeAggregate =>
+        aggBufferAttributes ++= agg.aggBufferAttributes
+        initialValues ++= agg.initialValues
+        updateExpressions ++= agg.updateExpressions
+        evaluateExpressions += agg.evaluateExpression
+      case agg: ImperativeAggregate =>
+        val offset = aggBufferAttributes.size
+        val imperative = BindReferences.bindReference(agg
+          .withNewInputAggBufferOffset(offset)
+          .withNewMutableAggBufferOffset(offset),
+          inputAttributes)
+        imperatives += imperative
+        aggBufferAttributes ++= imperative.aggBufferAttributes
+        val noOps = Seq.fill(imperative.aggBufferAttributes.size)(NoOp)
+        initialValues ++= noOps
+        updateExpressions ++= noOps
+        evaluateExpressions += imperative
+      case other =>
+        sys.error(s"Unsupported Aggregate Function: $f")
+    }
+
+    // Process the functions.
+    functions.foreach(addSize.andThen(addToProcessor))
+
+    // Create the projections.
+    val initialProjection = newMutableProjection(initialValues, Nil)()
+    val updateProjection = newMutableProjection(
+      updateExpressions,
+      aggBufferAttributes ++ inputAttributes)()
+    val evaluateProjection = newMutableProjection(
+      evaluateExpressions,
+      aggBufferAttributes)()
```

Comment:
I have added the bound ```ImperativeAggregate``` to the ```evaluateExpressions```, so there is (should be) nothing left to bind. The ```ImperativeAggregate``` is being bound on line https://github.com/hvanhovell/spark/blob/SPARK-8641-2/sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala#L753